shell/platform/fuchsia/flutter/flatland_connection.cc

```diff
@@ -51,18 +51,36 @@ void FlatlandConnection::Present() {
   acquire_fences_.clear();
 }
 
+// Note that this method is called from the UI thread.
 void FlatlandConnection::AwaitVsync(FireCallbackCallback callback) {
-  if (first_call) {
+  if (first_call_to_await_vsync_) {
     fml::TimePoint now = fml::TimePoint::Now();
     callback(now, now + kDefaultFlatlandPresentationInterval);
-    first_call = false;
+    first_call_to_await_vsync_ = false;
     return;
   }
-  fire_callback_ = callback;
+
+  {
+    std::scoped_lock<std::mutex> lock(threadsafe_state_.mutex_);
+    threadsafe_state_.fire_callback_ = callback;
+
+    if (threadsafe_state_.fire_callback_pending_) {
+      fml::TimePoint now = fml::TimePoint::Now();
+      // TODO(fxbug.dev/64201): Calculate correct frame times.
+      threadsafe_state_.fire_callback_(
+          now, now + kDefaultFlatlandPresentationInterval);
+      threadsafe_state_.fire_callback_ = nullptr;
+      threadsafe_state_.fire_callback_pending_ = false;
+    }
+  }
 }
 
+// Note that this method is called from the UI thread.
 void FlatlandConnection::AwaitVsyncForSecondaryCallback(
-    FireCallbackCallback callback) {}
+    FireCallbackCallback callback) {
+  fml::TimePoint now = fml::TimePoint::Now();
+  callback(now, now);
+}
 
 void FlatlandConnection::OnError(
```

Comment:
Above comments on which thread methods runs is very useful? Can we add the rest too, i.e. raster thread?

Reply:
Done